src/test/denialofservice_tests.cpp

```diff
@@ -106,25 +106,41 @@ BOOST_AUTO_TEST_CASE(outbound_slow_chain_eviction)
     peerman.FinalizeNode(dummyNode1);
 }
 
-static void AddRandomOutboundPeer(NodeId& id, std::vector<CNode*>& vNodes, PeerManager& peerLogic, ConnmanTestMsg& connman, ConnectionType connType)
+static void AddRandomOutboundPeer(NodeId& id, std::vector<CNode*>& vNodes, PeerManager& peerLogic, ConnmanTestMsg& connman, ConnectionType connType, bool onion_peer = false)
 {
-    CAddress addr(ip(g_insecure_rand_ctx.randbits(32)), NODE_NONE);
-    vNodes.emplace_back(new CNode{id++,
-                                  /*sock=*/nullptr,
-                                  addr,
-                                  /*nKeyedNetGroupIn=*/0,
-                                  /*nLocalHostNonceIn=*/0,
-                                  CAddress(),
-                                  /*addrNameIn=*/"",
-                                  connType,
-                                  /*inbound_onion=*/false});
-    CNode &node = *vNodes.back();
-    node.SetCommonVersion(PROTOCOL_VERSION);
-
-    peerLogic.InitializeNode(node, ServiceFlags(NODE_NETWORK | NODE_WITNESS));
-    node.fSuccessfullyConnected = true;
-
-    connman.AddTestNode(node);
+    auto routable = false;
+
+    while (!routable) {
```

Comment:
Instead of pushing the node to `vNodes` and removing it if it's not routable, I think you could do something like:
```cpp
static void AddRandomOutboundPeer(NodeId& id, std::vector<CNode*>& vNodes, PeerManager& peerLogic, ConnmanTestMsg& connman, ConnectionType connType, bool onion_peer = false)
{
    CAddress addr;

    if (onion_peer) {
        auto tor_addr{g_insecure_rand_ctx.randbytes(ADDR_TORV3_SIZE)};
        BOOST_REQUIRE(addr.SetSpecial(OnionToString(tor_addr)));
    }

    while (!addr.IsRoutable()) {
        addr = CAddress(ip(g_insecure_rand_ctx.randbits(32)), NODE_NONE);
    }

    vNodes.emplace_back(new CNode{id++,
            /*sock=*/nullptr,
            addr,
            /*nKeyedNetGroupIn=*/0,
            /*nLocalHostNonceIn=*/0,
            CAddress(),
            /*addrNameIn=*/"",
            connType,
            /*inbound_onion=*/false});
    CNode &node = *vNodes.back();

    node.SetCommonVersion(PROTOCOL_VERSION);
    peerLogic.InitializeNode(node, ServiceFlags(NODE_NETWORK | NODE_WITNESS));
    node.fSuccessfullyConnected = true;
    connman.AddTestNode(node);
}
```

1. You can check if the `addr` is routable directly, no need to do it in `CNode`.
2. Check `onion_peer` before trying to create a routable addr.